File: java/src/test/java/fr/sdecout/handson/rest/library/PostalCodeTest.java
package fr.sdecout.handson.rest.library;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;

public class PostalCodeTest {

    @Test
    void should_fail_to_initialize_from_value_with_less_than_5_digits() {
        var value = "1111";
        assertThat(value).hasSize(4);

        assertThatIllegalArgumentException()
                .isThrownBy(() -> new PostalCode(value))
                .withMessage("Postal code must contain 5 numeric digits");
    }

    @Test
    void should_fail_to_initialize_from_value_with_more_than_5_digits() {
        var value = "111111";
        assertThat(value).hasSize(6);

        assertThatIllegalArgumentException()
                .isThrownBy(() -> new PostalCode(value))
                .withMessage("Postal code must contain 5 numeric digits");
    }

    @Test
    void should_fail_to_initialize_from_value_including_non_numeric_digits() {
        var value = "1111a";
        assertThat(value).hasSize(5);

        assertThatIllegalArgumentException()
                .isThrownBy(() -> new PostalCode(value))
                .withMessage("Postal code must contain 5 numeric digits");
    }

    @Test
    void should_get_department_code() {
        assertThat(new PostalCode("28200").departmentCode()).isEqualTo("28");
    }

    @Test
    void should_display_as_string() {
        var value = "11111";
        assertThat(value).hasSize(5);

        assertThat(new PostalCode(value)).hasToString(value);
    }

}

File: kotlin/src/test/kotlin/fr/sdecout/handson/rest/library/PostalCodeTest.kt
package fr.sdecout.handson.rest.library

import fr.sdecout.handson.rest.library.PostalCode
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatIllegalArgumentException
import org.junit.jupiter.api.Test

class PostalCodeTest {

    @Test
    fun `should fail to initialize from value with less than 5 digits`() {
        val value = "1111"
        assertThat(value).hasSize(4)

        assertThatIllegalArgumentException()
            .isThrownBy { PostalCode(value) }
            .withMessage("Postal code must contain 5 numeric digits")
    }

    @Test
    fun `should fail to initialize from value with more than 5 digits`() {
        val value = "111111"
        assertThat(value).hasSize(6)

        assertThatIllegalArgumentException()
            .isThrownBy { PostalCode(value) }
            .withMessage("Postal code must contain 5 numeric digits")
    }

    @Test
    fun `should fail to initialize from value including non-numeric digits`() {
        val value = "1111a"
        assertThat(value).hasSize(5)

        assertThatIllegalArgumentException()
            .isThrownBy { PostalCode(value) }
            .withMessage("Postal code must contain 5 numeric digits")
    }

    @Test
    fun `should get department code`() {
        assertThat(PostalCode("28200").departmentCode).isEqualTo("28")
    }

    @Test
    fun `should display as string`() {
        val value = "11111"
        assertThat(value).hasSize(5)

        assertThat(PostalCode(value)).hasToString(value)
    }

}
